pkg/generator/vault/vault.go

```diff
@@ -0,0 +1,98 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vaultdynamic
+
+import (
+	"context"
+	"fmt"
+
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/client-go/kubernetes"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlcfg "sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/yaml"
+
+	genv1alpha1 "github.com/external-secrets/external-secrets/apis/generators/v1alpha1"
+	provider "github.com/external-secrets/external-secrets/pkg/provider/vault"
+)
+
+type Generator struct{}
+
+const (
+	errNoSpec      = "No config spec provided"
+	errParseSpec   = "Unable to parse spec: %w"
+	errVaultClient = "Unable to setup Vault client: %w"
+	errGetSecret   = "Unable to get dynamic secret: %w"
+)
+
+func (g *Generator) Generate(ctx context.Context, jsonSpec *apiextensions.JSON, kube client.Client, namespace string) (map[string][]byte, error) {
+	c := &provider.Connector{NewVaultClient: provider.NewVaultClient}
+
+	// controller-runtime/client does not support TokenRequest or other subresource APIs
+	// so we need to construct our own client and use it to fetch tokens
+	// (for Kubernetes service account token auth)
+	restCfg, err := ctrlcfg.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	clientset, err := kubernetes.NewForConfig(restCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return g.generate(c, ctx, jsonSpec, kube, clientset.CoreV1(), namespace)
+}
+
+func (g *Generator) generate(c *provider.Connector, ctx context.Context, jsonSpec *apiextensions.JSON, kube client.Client, corev1 typedcorev1.CoreV1Interface, namespace string) (map[string][]byte, error) {
+	if jsonSpec == nil {
+		return nil, fmt.Errorf(errNoSpec)
+	}
+	res, err := parseSpec(jsonSpec.Raw)
+	if err != nil {
+		return nil, fmt.Errorf(errParseSpec, err)
+	}
+	if res == nil || res.Spec.Vault == nil {
+		return nil, fmt.Errorf("No Vault config in spec")
+	}
+	cl, err := c.NewGeneratorClient(ctx, kube, corev1, res.Spec.Vault, namespace)
+	if err != nil {
+		return nil, fmt.Errorf(errVaultClient, err)
+	}
+
+	result, err := (*cl).Logical().ReadWithDataWithContext(ctx, res.Spec.Path, nil)
+	if err != nil {
+		return nil, err
+	}
+	if result == nil {
+		return nil, fmt.Errorf(errGetSecret, fmt.Errorf("Empty response from Vault"))
+	}
+
+	response := make(map[string][]byte)
```

Comment:
The response can be deeply nested, this should be a `map[string]interface{}`.

Reply:
I've tried to make this work, but the [Generator interface](https://github.com/external-secrets/external-secrets/blob/7e5fbb124b4cb24db7b7765bc061b7b1ff2abfda/apis/generators/v1alpha1/generator.go#L34) requires the `map[string][]byte` return type.

The current setup means nested fields are stored in the target `Secret` as their Go string representation (for example, `map[contents:read metadata:read]`). Perhaps I could improve that further by `JSON.Marshal`ling them?

That way, the current `Generate` interface type can be respected, and target users can parse the required values from JSON as needed (while top-level string values would be readily available to use).

Reply:
Looking at it :eyes: this sounds reasonable. It doesn't really make sense to change the interface, as the target Secret can only hold simple key/value pairs, not nested values. 

Nested values are pretty rarely used in vault AFAIK, so :+1:

Reply:
Thanks @moolen, I've pushed an update that makes use of (and extends) the function you mentioned (works for me for `github`-type dynamic secret in live environment).